src/crypto/CHIPCryptoPALOpenSSL.cpp

```diff
@@ -1665,5 +1670,76 @@ CHIP_ERROR ExtractPubkeyFromX509Cert(const ByteSpan & certificate, Crypto::P256P
     return err;
 }
 
+CHIP_ERROR ExtractAKIDFromX509Cert(const ByteSpan & certificate, MutableByteSpan & akid)
+{
+    CHIP_ERROR err                       = CHIP_NO_ERROR;
+    X509 * x509certificate               = nullptr;
+    const unsigned char * pCertificate   = certificate.data();
+    const unsigned char ** ppCertificate = &pCertificate;
+    const ASN1_OCTET_STRING * akidString = nullptr;
+
+    x509certificate = d2i_X509(NULL, ppCertificate, static_cast<long>(certificate.size()));
+    VerifyOrExit(x509certificate != nullptr, err = CHIP_ERROR_NO_MEMORY);
+
+    akidString = X509_get0_authority_key_id(x509certificate);
+
+    VerifyOrExit(akidString->length == static_cast<int>(akid.size()), err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+
+    memcpy(akid.data(), akidString->data, static_cast<size_t>(akidString->length));
+
+exit:
+    X509_free(x509certificate);
+
+    return err;
+}
+
+CHIP_ERROR ExtractVIDFromX509Cert(const ByteSpan & certificate, VendorId & vid)
+{
+    CHIP_ERROR err                     = CHIP_NO_ERROR;
+    X509 * x509certificate             = nullptr;
+    const unsigned char * pCertificate = certificate.data();
+    constexpr char vidNeedle[]         = "1.3.6.1.4.1.37244.2.1"; // Matter VID OID - taken from Spec
+    constexpr size_t vidNeedleSize     = sizeof(vidNeedle);
+    char buff[vidNeedleSize]           = { 0 };
+    X509_NAME * subject                = nullptr;
+    int x509EntryCountIdx              = 0;
+
+    vid = VendorId::NotSpecified;
+
+    x509certificate = d2i_X509(NULL, &pCertificate, static_cast<long>(certificate.size()));
+    VerifyOrExit(x509certificate != nullptr, err = CHIP_ERROR_NO_MEMORY);
+
+    subject = X509_get_subject_name(x509certificate);
+    VerifyOrExit(subject != nullptr, err = CHIP_ERROR_INTERNAL);
+
+    for (x509EntryCountIdx = 0; x509EntryCountIdx < X509_NAME_entry_count(subject); ++x509EntryCountIdx)
+    {
+        X509_NAME_ENTRY * name_entry = X509_NAME_get_entry(subject, x509EntryCountIdx);
+        VerifyOrExit(name_entry != nullptr, err = CHIP_ERROR_INTERNAL);
+        ASN1_OBJECT * object = X509_NAME_ENTRY_get_object(name_entry);
+        VerifyOrExit(object != nullptr, err = CHIP_ERROR_INTERNAL);
+        VerifyOrExit(OBJ_obj2txt(buff, sizeof(buff), object, 0) != 0, err = CHIP_ERROR_INTERNAL);
+
+        if (memcmp(vidNeedle, buff, vidNeedleSize) == 0)
```

Comment:
If vidNeedle and buff are null-terminated strings, why are we using memcmp instead of a string function?

Reply:
(I.e. strncmp)